ext/ddtrace_profiling_native_extension/linux_tid_from_pthread.c

```diff
@@ -0,0 +1,209 @@
+
+  struct tid_probe_info {
+    pid_t tid;
+    char buffer[STRUCT_PTHREAD_READ_SIZE];
+    bool buffer_read_success;
+  };
+
+  static void *collect_tid_probe_info(void *tid_probe_info);
+  static short find_tid_offset_in_buffer(struct tid_probe_info *tid_probe_info, short start_position);
+  static bool read_struct_pthread(pthread_t thread, void *struct_pthread_buffer);
+  static pid_t get_tid_from_buffer(void *struct_pthread_buffer, short offset);
+
+  short setup_linux_tid_from_pthread_offset(void) {
+    int error;
+    struct tid_probe_info reference_threads_probe_info[REFERENCE_THREADS_COUNT];
+
+    // We use the current thread as the first reference thread
+    collect_tid_probe_info(&reference_threads_probe_info[0]);
+    if (!reference_threads_probe_info[0].buffer_read_success) return -1;
+
+    // ...and then we create a few more to act as extra references
+    for (int i = 1; i < REFERENCE_THREADS_COUNT; i++) {
+      pthread_t reference_thread;
+      error = pthread_create(&reference_thread, NULL, collect_tid_probe_info, &reference_threads_probe_info[i]);
+      if (error) return -1;
+      error = pthread_join(reference_thread, NULL);
+      if (error || !reference_threads_probe_info[i].buffer_read_success) return -1;
+    }
+
+    short candidate_tid_offset = 0;
+
+    // We try to find an offset that contains the expected tid for all reference threads. If we get to the end of the
+    // buffer before finding such an offset, the `thread_offset` will be -1 and we'll report back an error.
+    while (candidate_tid_offset >= 0) {
+      bool all_offsets_match = true;
+
+      for (int i = 0; i < REFERENCE_THREADS_COUNT; i++) {
+        short thread_offset = find_tid_offset_in_buffer(&reference_threads_probe_info[i], candidate_tid_offset);
+
+        if (thread_offset != candidate_tid_offset) {
+          all_offsets_match = false;
+          candidate_tid_offset = thread_offset;
+        }
+      }
+
+      if (all_offsets_match) break;
+    }
+
+    return candidate_tid_offset;
+  }
+
+  pid_t linux_tid_from(pthread_t thread, short offset) {
+    char buffer[STRUCT_PTHREAD_READ_SIZE];
+    read_struct_pthread(thread, buffer);
+    return get_tid_from_buffer(buffer, offset);
+  }
+
+  static void *collect_tid_probe_info(void *tid_probe_info) {
+    struct tid_probe_info *result = (struct tid_probe_info *) tid_probe_info;
+
+    result->tid = ddtrace_gettid();
+    result->buffer_read_success = read_struct_pthread(pthread_self(), result->buffer);
+
+    return NULL;
+  }
+
+  bool read_safely(void *read_from_ptr, void *read_into_buffer, short buffer_size) {
+    struct iovec read_into = {.iov_base = read_into_buffer, .iov_len = buffer_size};
+    struct iovec read_from = {.iov_base = read_from_ptr, .iov_len = buffer_size};
+
+    int unused_flags = 0;
+    int number_of_iovecs = 1;
+
+    short num_read = process_vm_readv(ddtrace_gettid(), &read_into, number_of_iovecs, &read_from, number_of_iovecs, unused_flags);
```

Comment:
Ok so hear me out: what if we replaced all reads with `process_vm_readv`, so that segfaults can never exist ever again?!

Reply:
This is actually not that wild of an idea. I can't remember any reported segfaults from the profiler sampling code, but definitely using `process_vm_readv` would increase safety when reading VM data structures (in fact, we experimented with `process_vm_readv` for implementing at least one other profiler, but ended up taking a different approach).

I'd say the main disadvantage is that `process_vm_readv` can require a few more security permissions (see: it's disabled on CircleCI) thus it would complicate/restrict onboarding, which is a shame.